Merge the react-native-webrtc with the version 1.94.1 of upstream

filipi87 opened the pull request:
Bringing all the changes from the version 1.94.1 to our fork.

filipi87 commented:
Has already been validated that is working on IOs, still need to validate that is working on IOs.

regexident commented on ios/RCTWebRTC/VideoCaptureController.m:
```diff
@@ -4,32 +4,36 @@
-        _capturer = capturer;
-        _running = NO;
+        self.capturer = capturer;
+        self.running = NO;
 
         // Default to the front camera.
-        _usingFrontCamera = YES;
+        self.usingFrontCamera = YES;
 
-        _deviceId = constraints[@"deviceId"];
-        _width = [constraints[@"width"] intValue];
-        _height = [constraints[@"height"] intValue];
-        _fps = [constraints[@"frameRate"] intValue];
+        self.deviceId = constraints[@"deviceId"];
+        self.width = [constraints[@"width"] intValue];
+        self.height = [constraints[@"height"] intValue];
+        self.frameRate = [constraints[@"frameRate"] intValue];
```
General advise in ObjC is to not use properties in initializers (as those could be accessing uninitialized state).
But since their properties are synthesized the code should be fine.

regexident commented on ios/RCTWebRTC/VideoCaptureController.m:
```diff
@@ -45,97 +49,146 @@ -(instancetype)initWithCapturer:(RTCCameraVideoCapturer *)capturer
-    AVCaptureDevice *device;
-    if (_deviceId) {
-        device = [AVCaptureDevice deviceWithUniqueID:_deviceId];
+- (void)dealloc {
+    self.device = NULL;
+}
```
This is not necessary.

regexident reviewed (commented):
Just some minor nits. Nothing troubling though.

filipi87 commented:
I have validated the changes on both IOs and Android, now It is working on both after some adjustments. 
It would be important if you were able to see It working on them either.
Thanks a lot for your help!

regexident reviewed (approved):
LGTM!

mattieruth commented on src/MediaStreamTrack.js:
```diff
@@ -0,0 +1,118 @@
+        if (this.kind !== 'video') {
+            throw new Error('Only implemented for video tracks');
+        }
+        return WebRTCModule.mediaStreamTrackSwitchCamera(this.id);
```
i'm out of my element here but... in the merge conflict resolution it looks like this changes react's version which does not return here. did you mean to put the `return` back in?

filipi87 replied:
One of the changes that we have made at our version of react-native-webrtc, is to return when we are switching the camera, which is the new facingMode result. That is why we have this return.

mattieruth replied:
👍  just wanted to make sure it was intentional.

mattieruth commented on android/src/main/java/com/oney/WebRTCModule/WebRTCModule.java:
```diff
@@ -417,29 +415,30 @@ public void peerConnectionInit(ReadableMap configuration, int id) {
+        // This function _only_ gets called from WebRTCView, in the UI thread.
+        // Hence make sure we run this code in the executor or we run at the risk
+        // of being out of sync.
         try {
-            return ThreadUtils.runOnExecutorAndWait(() -> getStreamForReactTagAsync(streamReactTag));
-        } catch (Exception e) {
-            Log.d(TAG, "getStreamForReactTag() failed");
-            e.printStackTrace();
-        }
-        return null;
-    }
+            return (MediaStream) ThreadUtils.submitToExecutor((Callable<Object>) () -> {
+                MediaStream stream = localStreams.get(streamReactTag);
 
-    MediaStream getStreamForReactTagAsync(String streamReactTag) {
-        MediaStream stream = localStreams.get(streamReactTag);
-
-        if (stream == null) {
-            for (int i = 0, size = mPeerConnectionObservers.size(); i < size; i++) {
-                PeerConnectionObserver pco = mPeerConnectionObservers.valueAt(i);
-                stream = pco.remoteStreams.get(streamReactTag);
```
these changes came in from the second merge commit: https://github.com/daily-co/react-native-webrtc/pull/17/commits/1217b35e6a7be37bf0da08b5b68acb37d3040c70

but seem odd since you remove a whole function... can you expand on these changes?  again... i'm totally out of my element

filipi87 replied:
The WebRTCModule.java is one of the classes that we have changed, to support basically this features:
- Return a value from the method `mediaStreamTrackSwitchCamera`
- Creating a new method `mediaStreamTrackGetCameraFacingMode`
- Creating a new method `setDailyAudioMode`

When I did the merge, that class had a lot of changes between the version 1.89.4 and this version 1.94.1. So, for some reason, I didnt merge right those parts that you have described above. That is why I needed to create a new commit, to remove the parts that have been wrong merged. Those parts were not supposed to exist in this version 1.94.1.

You can see here the comparative between how is this branch now and the upstream from react-native, you will see that only have the new features that I have described:
https://github.com/react-native-webrtc/react-native-webrtc/compare/master...daily-co:main-synced-to-1.94.1#diff-6732b1c8f20f6bb8141dcabe3d92b5d92afe0cbc8ce806d3dd7d68bfe9cb26c3

mattieruth reviewed (approved):
lots of caveats around my approval since i have no familiarity here, but my questions have been answered :-) -- @kompfner's turn

![](https://media.giphy.com/media/48FhEMYGWji8/giphy.gif)

kompfner commented:
reviewing now, using https://github.com/react-native-webrtc/react-native-webrtc/compare/1.94.1...daily-co:main-synced-to-1.94.1 to validate that daily custom stuff ported over

kompfner commented:
hmm, did we lose [this bugfix](https://github.com/daily-co/react-native-webrtc/commit/f58e1a9b89777931f8891a839fc2989af0a121fe)? i don't see the change to `getStreamForReactTag` in WebRTCModule.java anymore...

do you happen to know if it's not needed anymore?

kompfner commented:
looks like there's been [some changes lately](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/master/types/react-native-webrtc/index.d.ts) to DefinitelyTyped's `index.d.ts` since we last synced our repo with upstream—maybe we should take this opportunity to also update it?

the bottom of our README-internal has notes on our slightly wonky TS setup. maybe we should add a note to the "how to sync with upstream changes" section mentioning that when we sync it's also a good time to remember to check for and pull in changes to the DefinitelyTyped typescript declaration?

kompfner commented:
Looks like you maybe introduced a number of accidental tab/space changes in VideoCaptureController.m?

kompfner commented on ios/RCTWebRTC/WebRTCModule+Daily.m:
```diff
@@ -126,27 +126,31 @@ - (NSString *)audioMode {
-  // after being reset by WebRTC's audio module (for example, after a Wifi -> LTE
-  // switch), so (re-)apply the currently chosen audio mode to the session.
-  [self applyAudioMode:self.audioMode toSession:audioSession];
+  dispatch_async(self.captureSessionQueue, ^{
```
hm, i'm afraid we might be sweeping a different problem under the rug by just dispatching audio session configuration changes onto a different background queue...

at the very least we should give it a new queue and not reuse the one being used by the no-op capture session (whose purpose is to ensure that the mic is turned on so the app stays running if backgrounded)

kompfner replied:
regardless of what we do here, let's make sure to test the heck out of react-native-daily-js's `setAudioMode()`, as well as background the app before it finishes connecting to the room, to ensure that the no-op-audio-recording-to-keep-the-app-alive mechanism still works.

sorry if this feedback is for previous work and not strictly this PR! maybe this feedback can be applied in a follow-up PR

filipi87 replied:
Hi Paul, As we have discussed, I have changed for this part of audio mode to use the same queue that the react-native-webrtc already uses. The webrtc already uses this queue internally, so this way I believe we have an unexpected side effect.

I have also taken the latest changes from the upstream of react-native-webrtc, that they have committed in the last two weeks. Updated our DefinedTypes and removed the method that we were not using anymore at ThreadUtils.

After that, I have repeated the tests, both on Android and IOs, and it looks like everything is still working fine. The only test that I have not been able to execute is with the bluetooth phone ( Still dont have one. =// ). So if you could help with that, it would be great. Actually, the more we test, the better. =)

filipi87 commented:
> one question: is the plan to force push this branch up and replace `main`, per our internal README instructions? i feel like there was a reason we had landed on that approach, rather than merging this branch in...something to do with the resulting git history?
> 
> if so, we should make sure to make a backup branch of current `main` called `main-pre-syncing-to-1.94.1` before we do the force push.

Yes, that is the idea. I have already created this backup branch, before starting the merge => main-pre-syncing-to-1.94.1

filipi87 commented:
> hmm, did we lose [this bugfix](https://github.com/daily-co/react-native-webrtc/commit/f58e1a9b89777931f8891a839fc2989af0a121fe)? i don't see the change to `getStreamForReactTag` in WebRTCModule.java anymore...
> 
> do you happen to know if it's not needed anymore?

I believe we dont need that anymore. Before, If you look the difference of our version 1.89 and theirs: https://github.com/react-native-webrtc/react-native-webrtc/compare/1.89.3...daily-co:main

They tried to access the local stream directly, and that is why I believe we have changed to retrieve from the Executor. But now, they are already doing that:

```
MediaStream getStreamForReactTag(String streamReactTag) {
        // This function _only_ gets called from WebRTCView, in the UI thread.
        // Hence make sure we run this code in the executor or we run at the risk
        // of being out of sync.
        try {
            return (MediaStream) ThreadUtils.submitToExecutor((Callable<Object>) () -> {
                MediaStream stream = localStreams.get(streamReactTag);

                if (stream != null) {
                    return stream;
                }

                for (int i = 0, size = mPeerConnectionObservers.size(); i < size; i++) {
                    PeerConnectionObserver pco = mPeerConnectionObservers.valueAt(i);
                    stream = pco.remoteStreams.get(streamReactTag);
                    if (stream != null) {
                        return stream;
                    }
                }

                return null;
            }).get();
        } catch (ExecutionException | InterruptedException e) {
            return null;
        }
    }
```

filipi87 commented:
> looks like there's been [some changes lately](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/master/types/react-native-webrtc/index.d.ts) to DefinitelyTyped's `index.d.ts` since we last synced our repo with upstream—maybe we should take this opportunity to also update it?
> 
> the bottom of our README-internal has notes on our slightly wonky TS setup. maybe we should add a note to the "how to sync with upstream changes" section mentioning that when we sync it's also a good time to remember to check for and pull in changes to the DefinitelyTyped typescript declaration?

Sure, I will pull this last change from the upstream. Looks like they have only changed a type, dont think will bring any kind of side effects: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/57353/files

filipi87 commented:
> Looks like you maybe introduced a number of accidental tab/space changes in VideoCaptureController.m?

I believe my IDE is configured to apply this automatically.

kompfner commented:
Looks like we ended up synced *past* the upstream 1.94.1 release. Let's try to sync strictly with releases, since:
* those are (theoretically, at least) the most stable/well-tested
* it makes it easier for us to keep track of where our fork is synced to

filipi87 commented:
> Looks like we ended up synced _past_ the upstream 1.94.1 release. Let's try to sync strictly with releases, since:
> 
> * those are (theoretically, at least) the most stable/well-tested
> * it makes it easier for us to keep track of where our fork is synced to

Have just fixed that, have rebased for before this commit, and apply the other commits later.

kompfner commented:
Shoot, in testing this branch I'm seeing a resurgence of a lot of the kinds of familiar probably-thread-starving-related instability I had (with difficulty) reached an uneasy truce with over a few horrible days in May, by jiggling things around and then backing away slowly:
* first with [this commit](https://github.com/daily-co/react-native-webrtc/commit/fbcaae967998b1dd7eeb13cc8c1c9b92f6340768)
* then [this one](https://github.com/daily-co/react-native-webrtc/commit/726e8b20b16d134a7faa032a0ff4c53a0bd628ef)
* and finally [this one](https://github.com/daily-co/react-native-webrtc/commit/580124be0ea91ed9a4c3b92328135078a053aa6e) (where I **finally** thought I had a relatively clean solution)

See commit messages for descriptions.

filipi87 commented:
> 

I have spent today testing the IOs app, and the only way I was able to reproduce the connection issues, is when I connect to the meeting, and really fast put the app to background, then I receive the wss disconnects and errors. But I have been able to do that only 02 times so far. But as you know, I do not have a bluetooth phone to test those scenarios. 

However, that made me wonder, the problem has just happened, because probably our captureSession was not ready. My perception here is that It takes something between 500ms and 1s until I am able to put the app to background, if we do before that, It is not going to work.

And the only purpose of creating this captureSession, is the meeting remains active, if the user puts our app to background or blocks the screen before he has actually joined the conference. The time to join a conference is about 2s +-. 

So what I was thinking is, is It worth risking some unknown side effects, just to prevent the user to wait like 1 second more ? 

Besides that, the other that we change is the setAudioMode, so I investigate how the react-native-incall-manager does that. 

https://github.com/react-native-webrtc/react-native-incall-manager/blob/master/ios/RNInCallManager/RNInCallManager.m

I found that interesting comment there:
// --- WebRTC native code will change audio mode automatically when established.
// --- It would have some race condition if we change audio mode with webrtc at the same time.
// --- So we should not change audio mode as possible as we can. Only when default video call which wants to force speaker off.
// --- audio: only override speaker on/off; video: should change category if needed and handle proximity sensor. ( because default proximity is off when video call )

And when they change the audio mode, they do of a different way than we are currently doing.

I have changed our way to be similar to the way that they are doing, only applying changes for the mode and the category, instead of changing the whole setConfiguration that we were doing before.

I have created a fork, `main-synced-to-1.94.1-ios`, so we can make some tests with those changes I have describede above. So far, for me everything is working fine.

This is the only file that I have changed so far:
https://github.com/daily-co/react-native-webrtc/blob/main-synced-to-1.94.1-ios/ios/RCTWebRTC/WebRTCModule%2BDaily.m

kompfner commented:
Really ran things through the gamut with your latest change, per the test plan laid out in [our discussion](https://www.notion.so/dailyco/iOS-issues-when-syncing-react-native-webrtc-with-upstream-version-1-94-1-f39c105849e844428e255fa2fc02e3f7#d0c6b0952d344373b4a4b80fdaac7957). Everything I tested seems to work quite well now!! 🥳  Nice work!

kompfner commented:
The only remaining piece of feedback is that I think we could possibly simplify the code a bit, per [Apple's recommendation](https://developer.apple.com/documentation/avfaudio/avaudiosession/1616614-setmode) that we avoid doing `setMode:` and `setCategory:` as separate calls—there's an API that will let us do it all at once. In RNInCallManager code you were inspired by, I suspect they are separate calls since they have a need to change each one independently, which I don't think we need.

I'll quickly test this simplification—if it seems to work as well, would you have any objection to including it here?

kompfner commented:
> The only remaining piece of feedback is that I think we could possibly simplify the code a bit, per [Apple's recommendation](https://developer.apple.com/documentation/avfaudio/avaudiosession/1616614-setmode) that we avoid doing `setMode:` and `setCategory:` as separate calls—there's an API that will let us do it all at once. In RNInCallManager code you were inspired by, I suspect they are separate calls since they have a need to change each one independently, which I don't think we need.
> 
> I'll quickly test this simplification—if it seems to work as well, would you have any objection to including it here?

It kind of seems like the simplification introduces a bit of instability—let's not risk it. I _think_ it's maybe due to reaching into `RTCAudioSession`'s `session` property and getting to the underlying `AVAudioSession` to use the desired method rather than using the `RTCAudioSession` proxy methods 🤷 . Either way, let's keep what's in this PR now and not mess with it. _Backs away slowly_.

kompfner commented:
^ pardon me: made minor indentation fixes in order to match what was there
